Unshorten url

antonydp opened the pull request:
Utils class to bypass shortner URLs
Based on https://github.com/kodiondemand/addon/blob/master/lib/unshortenit.py
Still a work in progress.
Since there's still a lot of work to be done, I would like to hear some opinions as to whether this can be useful to the project

antonydp commented:
Since there's still a lot of work to be done, I would like to hear some opinions if this could be helpful to the project

ArjixWasTaken commented:
In [this](https://github.com/LagradOst/CloudStream-3/pull/1172/files#diff-363538209598c6be246664b79351cf92a714539045615befb5670b37fe28e4ebR37-R118) function please use an `if else` chain

Like
```kotlin
if (false) { /* do smth */ }
else if (true) { /* do smth else */ }
```

ArjixWasTaken commented:
And please use `camelCase` instead of `snake_case`.

ArjixWasTaken commented:
> Since there's still a lot of work to be done, I would like to hear some opinions as to whether this can be useful to the project

Well, if it is guaranteed to work, then I bet it will be useful, in the past when we needed to bypass a shrinked url we just implemented the bypass on a provider basis.

For example [here](https://github.com/LagradOst/CloudStream-3/blob/master/app/src/main/java/com/lagradost/cloudstream3/animeproviders/AnimePaheProvider.kt#L385-L451) we bypass adfly.
Now that bypass may be outdated, who knows, the provider has been dead for like a year now, so it is unused.

antonydp commented:
I've rewritten the snake_case in camelCase as suggested and used a loop over a list instead of the if chain.
I've removed the not yet coded shortner, witch can be found here in the repository i've been taking inspiration: 
https://github.com/kodiondemand/addon/blob/master/lib/unshortenit.py 
I think it can be merged right now as we can keep adding useful url shortner as we go.

LagradOst commented on app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt:
```diff
@@ -0,0 +1,161 @@
+    val uprotRegex = """uprot\.net"""
+    val davisonbarkerRegex = """davisonbarker\.pro|lowrihouston\.pro"""
+
+    val shortList = listOf(
```
make this a companion object as well as the regex links, as you dont want to run unnecessary code for every unshorten.

antonydp replied:
is it ok the way I did it?

LagradOst replied:
yes

LagradOst commented on app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt:
```diff
@@ -0,0 +1,161 @@
+            r = app.get(uri, allowRedirects = true)
+            uri = r.url
+            val link =
+                Regex("<iframe[^<>]*src=\\'([^'>]*)\\'[^<>]*>").find(r.text)?.value ?:
```
the code will be slightly faster if you move regex to companion

antonydp replied:
I don't really understand what should I do, sorry

LagradOst commented on app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt:
```diff
@@ -0,0 +1,161 @@
+package com.lagradost.cloudstream3.utils
```
impo this should be an object and not just code

antonydp replied:
that's not entirely clear to me, sorry. could you explain a little bit more?

LagradOst replied:
```kotlin
object ShortUrl { 
[code here]
}
```
so you have to prefix it with ShortUrl.isShortLink when calling it

antonydp replied:
for both isShortLink and unshorten?

antonydp replied:
I wrote the ShortLink class, is it ok like that?

LagradOst commented on app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt:
```diff
@@ -0,0 +1,161 @@
+    return uri
+}
+fun unshortenLinksafe(uri:String) : String {
+    return Base64.decode(uri.split("?url=").last().toByteArray(), Base64.DEFAULT).decodeToString()
```
use mainapis base64 decode

antonydp replied:
done!

LagradOst reviewed (commented):
👍 just minor code changes, mainly that it should be an object and stuff should be computed once (like regex and shortList)

antonydp commented on app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt:
```diff
@@ -0,0 +1,161 @@
+        function
+    )
+}
+suspend fun unshorten(uri: String, type: String?): String {
```
yes!

antonydp commented on app/src/main/java/com/lagradost/cloudstream3/utils/UnshortenUrl.kt:
```diff
@@ -0,0 +1,161 @@
+import java.net.URI
+import java.net.URLDecoder
+
+//Code heavily based on unshortenit.py form kodiondemand /addon
```
I think I've done it.

LagradOst commented:
Very nice, do you want this merged now or do you want to add smth more?

antonydp commented:
I think it can be merged now, I'll try to add some more shorteners later on
